### ISSUE TITLE
[Proposal]: breaking point for C# Primary Constructor

### ISSUE THREAD

**Issue by ndepend** [dotnet/roslyn#67796]
I was investigating C# 12 `class` and `struct` primary constructors and figured out that no breakpoint can be set, unless a parameter is used within a property initialization. It can certainly be useful to break upon a primary constructor, the same way it is useful to be able to break upon a property `get;` or `set;`:
![C# 12 class and struct Primary Constructor and breakpoint](https://user-images.githubusercontent.com/511445/231777497-08b92683-39f9-4c0a-b92d-d3671e735991.png)

**Comment by tmat** [comment]
There are 3 issues here:
1) The compiler is not generating sequence points correctly for primary constructors for non-records. It works well for records.

Repro:

```C#
var s = new S(1, 2);
var p = new P(1, 2);
Console.WriteLine();

record struct S(int A, int B);
struct P(int A, int B) { }
```

Run and keep pressing `F11` to step into all debuggable code:

#### 1
![image](https://user-images.githubusercontent.com/41759/231823973-003a9532-66c3-408c-9504-c28b1217c485.png)
#### 2
![image](https://user-images.githubusercontent.com/41759/231824032-51ef33b0-eda7-48da-882d-cb2791a60d4c.png)
#### 3
![image](https://user-images.githubusercontent.com/41759/231824068-1fd34c77-97d6-4bc5-80f0-d55323ff6658.png)
#### 4
![image](https://user-images.githubusercontent.com/41759/231824107-e4f67e65-ec61-4a0c-a573-1a3c0ab62c54.png)

Compiler issue - missing sequence points (initializer of P is skipped):
#### 5
![image](https://user-images.githubusercontent.com/41759/231824135-dce90380-35cb-495a-b1c7-3763f7504859.png)
#### 6
![image](https://user-images.githubusercontent.com/41759/231824283-a5dc6c1f-56d7-4465-ab1a-ced133baf682.png)

2) The IDE is not allowing to place breakpoints for primary constructors (both records and non-records).

Repro: place cursor on `int A` in `S` or `P` and try to place a breakpoint (`F9`).

3) Not sure why, but in step 2 and 3 the grey highlight should be on `var s = new S(1, 2);`, not on `var p = new P(1, 2);`. Likely because of sequence points emitted around these statements, but it could also be debugger issue.

**Comment by tmat** [comment]
@AlekseyTs

**Comment by AlekseyTs** [comment]
> It works well for records.

I wouldn't say that. It looks to me that the issue is not specific to a particular flavor of a primary constructor, record or not. If a primary constructor doesn't have any user code, no sequence points generated for its body. For example, debugger doesn't stop for this record.
```
        record struct S(int A, int B)
        {
            public int A;
            public int B;
        }
```

Bottom line, this doesn't look like a new issue to me.

**Comment by ndepend** [comment]
@tmat indeed I didn't notice one can **step into** record primary constructor parameter declaration.
![image](https://user-images.githubusercontent.com/511445/231945118-b454c774-4e17-4a36-9ad1-a6f42ff6dd16.png)

Issue 1) No breaking point can be set with F9 in this context:
![image](https://user-images.githubusercontent.com/511445/231945173-652c9472-1b22-4939-bd74-f99b9d8a7b7b.png)

Issue 2) I cannot step-into with F11 in this context:
![image](https://user-images.githubusercontent.com/511445/231945303-9e5e4966-ec05-4695-b2eb-b6a4a77185fc.png)

Issue 3) No breaking point can be set with F9 in this context:
![image](https://user-images.githubusercontent.com/511445/231945368-e35114da-fd4f-4336-b97c-69467be5d582.png)

**Comment by egvijayanand** [comment]
In my case, the object instantiation would be from the DI container, so it is necessary to place a breakpoint at the type definition level to inspect it during the debugging process.

Otherwise, I need to traverse through a hierarchy of instances to start from the invocation point (Application -> View -> ViewModel). This is just for the sample created. A real-world example would be much more complex. As we assess this feature to be useful esp for Services where the constructor parameter just gets assigned to the read-only field for its lifetime.

And most of the time, the parameter would be used in another command, which is independent of instantiation, unlike the property where the value would be set instantly.

So need an option that addresses this issue before this feature hits GA.

**Comment by tmat** [comment]
Breakpoints in primary constructors are implemented in VS 17.8 Preview 1: https://github.com/dotnet/roslyn/pull/68765

**Comment by egvijayanand** [comment]
> Breakpoints in primary constructors are implemented in VS 17.8 Preview 1

Glad that it's getting addressed. Clarify whether v17.8 is planned to release before/alongside .NET 8 GA.
